pkg/search/controller.go

```diff
@@ -0,0 +1,414 @@
+}
+
+// handleErr checks if an error happened and makes sure we will retry later.
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	klog.Errorf("Error cache memeber cluster %v, %v", key, err)
+	c.queue.AddRateLimited(key)
+}
+
+// doCacheCluster processes the resourceRegistry object
+// TODO: update status
+func (c *Controller) doCacheCluster(cluster string) error {
+	// STEP0: ungregister cluster, stop all informers for a specific cluster, and delete the cluster from informer managers.
+	v, ok := c.clusterRegistry.Load(cluster)
+	if !ok {
+		klog.Infof("Cluster %s is not registered", cluster)
+		return nil
+	}
+
+	cr := v.(clusterRegistry)
+	if cr.unregistry() {
+		klog.Infof("try to stop cluster informer %s", cluster)
+		c.InformerManager.Stop(cluster)
+		return nil
+	}
+
+	// STEP1: deleted cluster, stop all informers for a specific cluster, and delete the cluster from informer managers.
+	cls, err := c.clusterLister.Get(cluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("try to stop cluster informer %s", cluster)
+			c.InformerManager.Stop(cluster)
+			return nil
+		}
+		return err
+	}
+
+	if !cls.DeletionTimestamp.IsZero() {
+		klog.Infof("try to stop cluster informer %s", cluster)
+		c.InformerManager.Stop(cluster)
+		return nil
+	}
+
+	// STEP2: added/updated cluster, builds an informer manager for a specific cluster.
+	if !c.InformerManager.IsManagerExist(cluster) {
+		klog.Info("try to build informer manager for cluster ", cluster)
+		controlPlaneClient := gclient.NewForConfigOrDie(c.restConfig)
+
+		clusterDynamicClient, err := util.NewClusterDynamicClientSet(cluster, controlPlaneClient)
+		if err != nil {
+			return err
+		}
+		_ = c.InformerManager.ForCluster(cluster, clusterDynamicClient.DynamicClientSet, 0)
+	}
+
+	if c.resourceHandler != nil {
+		sci := c.InformerManager.GetSingleClusterManager(cluster)
+		for gvr := range cr.resources {
+			klog.Infof("try to start informer for %s, %v", cluster, gvr)
+			// TODO: gvr exists check
+			sci.ForResource(gvr, c.resourceHandler)
+		}
+		sci.Start()
+		_ = sci.WaitForCacheSync()
+	}
+
+	return nil
+}
+
+// addResourceRegistry adds a resourceRegistry object to the queue
+func (c *Controller) addResoruceRegistry(obj interface{}) {
+	rr := obj.(*v1alpha1.ResourceRegistry)
+	resources := c.getResources(rr)
+
+	for _, cluster := range c.getClusters(rr) {
+		v, _ := c.clusterRegistry.LoadOrStore(cluster, clusterRegistry{
+			resources:  make(map[schema.GroupVersionResource]struct{}),
+			registries: make(map[string]struct{})})
+		cr := v.(clusterRegistry)
+
+		for _, r := range resources {
+			cr.resources[r] = struct{}{}
+		}
+		cr.registries[rr.GetName()] = struct{}{}
+		c.clusterRegistry.Store(cluster, cr)
+
+		c.queue.Add(cluster)
+	}
+}
+
+// updateResourceRegistry updates a resourceRegistry object in the queue
+func (c *Controller) updateResourceRegistry(oldObj, newObj interface{}) {
```

Comment:
@RainbowMango The case of deleting some watched resources (like pods) when modifying the ResourceRegistry has not been handled

Reply:
@RainbowMango I didn't find a way to stop a specific resource in SingleClusterInformerManager.